File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.java
import java.util.*;

class Solution {
    public int solution(String[][] book_time) {
        Arrays.sort(book_time, (a, b) -> timeToInt(a[0]) - timeToInt(b[0]));
        
        // 종료 시간 저장
        PriorityQueue<Integer> pq = new PriorityQueue<>();

        for (String[] time : book_time) {
            int start = timeToInt(time[0]);
            int end = timeToInt(time[1]) + 10;

            if (!pq.isEmpty() && pq.peek() <= start) {
                pq.poll(); // 방 재사용 가능
            }

            pq.offer(end); // 새로 배정되거나 재사용된 방의 종료 시각
        }

        return pq.size();
    }

    private int timeToInt(String time) {
        String[] parts = time.split(":");
        return Integer.parseInt(parts[0]) * 60 + Integer.parseInt(parts[1]);
    }
}

File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.py
import heapq

def to_min(t):
    h, m = map(int, t.split(":"))
    return h * 60 + m

def solution(book_time):
    times = []
    for s, e in book_time:
        start = to_min(s)
        end = to_min(e) + 10
        times.append((start, end))
    times.sort() ## start 기준 정렬

    ## min-heap : 각 방이 언제까지 사용 불가인지 (다시 사용 가능한 시각)
    heap = []

    for start, end in times:
        if heap and heap[0] <= start:
            # 가장 빨리 비는 방이 이번 예약 시작 전에 비면 방 재사용
            heapq.heappop(heap)
        # 이번 예약으로 방(새거나 재사용한거)의 새 end 시간 push
        heapq.heappush(heap, end)

    return len(heap)
